src/Chainweb/Mempool/InMem.hs

```diff
@@ -340,59 +372,118 @@ reaperThread cfg dataLock restore = forever $ do
+    lock <- readMVar lockMVar
+    let lastParentRef = _inmemLastNewBlockParent lock
+    return $ MempoolBackend
+      { mempoolTxConfig = tcfg
+      , mempoolBlockGasLimit = blockSizeLimit
+      , mempoolLastNewBlockParent = lastParentRef
+      , mempoolProcessFork = processFork
+      , mempoolMember = member
+      , mempoolLookup = lookup
+      , mempoolInsert = insert
+      , mempoolGetBlock = getBlock
+      , mempoolMarkValidated = markValidated
+      , mempoolMarkConfirmed = markConfirmed
+      , mempoolReintroduce = reintroduce
+      , mempoolGetPendingTransactions = getPending
+      , mempoolSubscribe = subscribe
+      , mempoolShutdown = shutdown
+      , mempoolClear = clear
+      }
   where
-    member = memberInMem lock
-    lookup = lookupInMem lock
-    insert = insertInMem broadcaster cfg lock
-    getBlock = getBlockInMem cfg lock
-    markValidated = markValidatedInMem cfg lock
-    markConfirmed = markConfirmedInMem lock
-    reintroduce = reintroduceInMem broadcaster cfg lock
-    getPending = getPendingInMem cfg lock
+    member = memberInMem lockMVar
+    lookup = lookupInMem lockMVar
+    insert = insertInMem broadcaster cfg lockMVar
+    getBlock = getBlockInMem cfg lockMVar
+    markValidated = markValidatedInMem cfg lockMVar
+    markConfirmed = markConfirmedInMem lockMVar
+    processFork = processForkInMem lockMVar blockHeaderDb payloadStore
+    reintroduce = reintroduceInMem broadcaster cfg lockMVar
+    getPending = getPendingInMem cfg lockMVar
     subscribe = subscribeInMem broadcaster
     shutdown = shutdownInMem broadcaster
-    clear = clearInMem lock
+    clear = clearInMem lockMVar
 
+------------------------------------------------------------------------------
+inMemPayloadLookup
+    :: forall cas . PayloadCas cas
+    => Maybe (PayloadDb cas)
+    -> BlockHeader
+    -> IO (Set ChainwebTransaction)
+inMemPayloadLookup payloadStore bh =
+    case payloadStore of
+        Nothing -> return mempty
+        Just s -> do
+            pwo <- casLookupM s (_blockPayloadHash bh)
+            MPCon.chainwebTxsFromPWO pwo
+  where
+    casLookupM s h = do
+        x <- casLookup s h
+        case x of
+            Nothing -> throwIO $ PayloadNotFoundException h
+            Just pwo -> return pwo
 
 ------------------------------------------------------------------------------
+processForkInMem :: PayloadCas cas
+                 => MVar (InMemoryMempoolData t)
+                 -> BlockHeaderDb
+                 -> Maybe (PayloadDb cas)
+                 -> LogFunction
+                 -> BlockHeader
+                 -> IO (Vector ChainwebTransaction)
+processForkInMem lock blockHeaderDb payloadStore logFun newHeader = do
+    theData <- readMVar lock
+    -- convert: Maybe (IORef BlockHeader) -> Maybe BlockHeader
+    lastHeader <- traverse readIORef (_inmemLastNewBlockParent theData)
+
+    MPCon.processFork logFun blockHeaderDb newHeader lastHeader
```

Comment:
...and this proves you don't need any specific knowledge of mempool internals to implement this, since the only field you're pulling out of it is the new one.